pyomo/solvers/plugins/solvers/SAS.py

```diff
@@ -0,0 +1,810 @@
+                + ";"
+            )
+        else:
+            return ""
+
+    def sas_version(self):
+        return self._sasver
+
+    def _apply_solver(self):
+        """ "Prepare the options and run the solver. Then store the data to be returned."""
+        logger.debug("Running SAS")
+
+        # Set return code to issue an error if we get interrupted
+        self._rc = -1
+
+        # Figure out if the problem has integer variables
+        with_opt = self.options.pop("with", None)
+        if with_opt == "lp":
+            proc = "OPTLP"
+        elif with_opt == "milp":
+            proc = "OPTMILP"
+        else:
+            # Check if there are integer variables, this might be slow
+            proc = "OPTMILP" if self._has_integer_variables() else "OPTLP"
+
+        # Remove CAS options in case they were specified
+        for opt in CAS_OPTION_NAMES:
+            self.options.pop(opt, None)
+
+        # Get the rootnode options
+        decomp_str = self._create_statement_str("decomp")
+        decompmaster_str = self._create_statement_str("decompmaster")
+        decompmasterip_str = self._create_statement_str("decompmasterip")
+        decompsubprob_str = self._create_statement_str("decompsubprob")
+        rootnode_str = self._create_statement_str("rootnode")
+
+        # Get a unique identifier, always use the same with different prefixes
+        unique = uuid.uuid4().hex[:16]
+
+        # Create unique filename for output datasets
+        primalout_dataset_name = "pout" + unique
+        dualout_dataset_name = "dout" + unique
+        primalin_dataset_name = None
+
+        # Handle warmstart
+        warmstart_str = ""
+        if self.warmstart_flag:
+            # Set the warmstart basis option
+            primalin_dataset_name = "pin" + unique
+            if proc != "OPTLP":
+                warmstart_str = """
+                                proc import datafile='{primalin}'
+                                    out={primalin_dataset_name}
+                                    dbms=csv
+                                    replace;
+                                    getnames=yes;
+                                    run;
+                                """.format(
+                    primalin=self._warm_start_file_name,
+                    primalin_dataset_name=primalin_dataset_name,
+                )
+                self.options["primalin"] = primalin_dataset_name
+
+        # Convert options to string
+        opt_str = " ".join(
+            option + "=" + str(value) for option, value in self.options.items()
+        )
+
+        # Set some SAS options to make the log more clean
+        sas_options = "option notes nonumber nodate nosource pagesize=max;"
+
+        # Get the current SAS session, submit the code and return the results
+        sas = self._sas_session
+        if sas == None:
+            sas = self._sas_session = self._sas.SASsession()
```

Comment:
Request here: I'd like to build in the capability for users to tell us where their `sascfg_personal.py` file is located. Can you please add a configuration option called `config_file` or something like that that, if not `None`, gets passed to the creation of the session?

Reply:
Okay this is now actually important because, surprise, we can run `saspy` on our machines now!

Reply:
One question about this: Do you want the cfgfile to be included in the options dictionary of the solve function or is there a way to include something like this in the SolverFactory constructor?